### PR TITLE
Vary Weighted Pool weights and swap fees

### LINKED ISSUE #813: Fuzz different pool settings in liquidity approximation
Liquidity approximation tests can be easily extended for new pool types, but each pool type can also have certain settings that are worth fuzzing. We might want to fuzz pool weights for weighted pools for example.

### PR DESCRIPTION
**Pull request by EndymionJkb**
# Description

In the direction of #813, this runs the basic Weighted Pool tests while fuzzing the weights between the minimum and maximum (also varying the swap fee for unbalanced adds).

It doesn't resolve #813, since that should really be another contract derived from LiquidityApproximationTest... but that test really deeply assumes equal weights/equal value, so it would be a fair bit of work to generalize it. The standard tests are much easier, and are a good first step. Perhaps the Mock introduced here can be reused there.

The fact that weights are immutable after deployment is really annoying for tests. To avoid deploying a new contract on every fuzz iteration, this PR introduces a WeightedPool mock that supports two tokens and has mutable weights. So you deploy it once at the beginning, and just set the weights on each iteration.

Since each test would change the state (balances, etc.), it uses snapshots to reset the state after each test runs. Both the weights and swap fee percentage are fuzzed; the swap fee percentage is only used in the last test.

## Type of change

- [ ] Bug fix 
- [ ] New feature 
- [ ] Breaking change 
- [ ] Dependency changes
- [ ] Code refactor / cleanup
- [ ] Optimization: [ ] gas / [ ] bytecode
- [ ] Documentation or wording changes
- [X] Other

## Checklist:

- [X] The diff is legible and has no extraneous changes
- [X] Complex code has been commented, including external interfaces
- [ ] Tests have 100% code coverage
- [X] The base branch is either `main`, or there's a description of how to merge

### CODE REVIEW

File: pkg/pool-weighted/test/foundry/WeightedPoolLimits.t.sol [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+    function _testInitialize() internal view {
+        // Tokens are transferred from lp.
+        assertEq(startingBalances[usdcIdx] - usdc.balanceOf(lp), amountsIn[usdcIdx], "LP: Wrong USDC balance");
+        assertEq(startingBalances[daiIdx] - dai.balanceOf(lp), amountsIn[daiIdx], "LP: Wrong DAI balance");
+
+        // Tokens are stored in the Vault.
+        assertEq(usdc.balanceOf(address(vault)), amountsIn[usdcIdx], "Vault: Wrong USDC balance");
+        assertEq(dai.balanceOf(address(vault)), amountsIn[daiIdx], "Vault: Wrong DAI balance");
+
+        // Tokens are deposited to the pool
```
Comment (joaobrunoah) [inline pkg/pool-weighted/test/foundry/WeightedPoolLimits.t.sol:179]:
```suggestion
        // Tokens are deposited to the pool.
```

File: pkg/pool-weighted/test/foundry/WeightedPoolLimits.t.sol [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+        vm.prank(bob);
+        bptAmountOut = router.addLiquidityUnbalanced(
+            address(pool),
+            newAmountsIn,
+            expectedBptAmountOut - DELTA,
+            false,
+            bytes("")
+        );
+
+        // Tokens are transferred from Bob
```
Comment (joaobrunoah) [inline pkg/pool-weighted/test/foundry/WeightedPoolLimits.t.sol:206]:
```suggestion
        // Tokens are transferred from Bob.
```

File: pkg/pool-weighted/test/foundry/WeightedPoolLimits.t.sol [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+        assertEq(initialBalances[usdcIdx] - usdc.balanceOf(bob), newAmountsIn[usdcIdx], "LP: Wrong USDC balance");
+        assertEq(initialBalances[daiIdx] - dai.balanceOf(bob), newAmountsIn[daiIdx], "LP: Wrong DAI balance");
+
+        expectedBalances[daiIdx] = amountsIn[daiIdx] + TOKEN_AMOUNT;
+        expectedBalances[usdcIdx] = amountsIn[usdcIdx] + TOKEN_AMOUNT;
+
+        // Tokens are stored in the Vault
```
Comment (joaobrunoah) [inline pkg/pool-weighted/test/foundry/WeightedPoolLimits.t.sol:213]:
```suggestion
        // Tokens are stored in the Vault.
```

File: pkg/pool-weighted/test/foundry/WeightedPoolLimits.t.sol [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+        assertEq(usdc.balanceOf(address(vault)), expectedBalances[usdcIdx], "Vault: Wrong USDC balance");
+        assertEq(dai.balanceOf(address(vault)), expectedBalances[daiIdx], "Vault: Wrong DAI balance");
+
+        // Tokens are deposited to the pool
```
Comment (joaobrunoah) [inline pkg/pool-weighted/test/foundry/WeightedPoolLimits.t.sol:217]:
```suggestion
        // Tokens are deposited to the pool.
```

File: pkg/pool-weighted/test/foundry/WeightedPoolLimits.t.sol [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+        (, , uint256[] memory balances, ) = vault.getPoolTokenInfo(address(weightedPool));
+        assertEq(balances[daiIdx], expectedBalances[daiIdx], "Pool: Wrong DAI balance");
+        assertEq(balances[usdcIdx], expectedBalances[usdcIdx], "Pool: Wrong USDC balance");
+
+        // should mint correct amount of BPT tokens
```
Comment (joaobrunoah) [inline pkg/pool-weighted/test/foundry/WeightedPoolLimits.t.sol:222]:
```suggestion
        // Should mint correct amount of BPT tokens.
```

File: pkg/pool-weighted/test/foundry/WeightedPoolLimits.t.sol [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
+            address(weightedPool),
+            bptAmountIn,
+            [limit, limit].toMemoryArray(),
+            false,
+            bytes("")
+        );
+
+        vm.stopPrank();
+
+        // Tokens are transferred to Bob
```
Comment (joaobrunoah) [inline pkg/pool-weighted/test/foundry/WeightedPoolLimits.t.sol:250]:
```suggestion
        // Tokens are transferred to Bob.
```

File: pkg/pool-weighted/test/foundry/WeightedPoolLimits.t.sol [hunk 7]
```diff
@@ -0,0 +1,383 @@
...
+            "Vault: Wrong USDC balance"
+        );
+        assertApproxEqAbs(dai.balanceOf(address(vault)), expectedBalances[daiIdx], DELTA, "Vault: Wrong DAI balance");
+
+        // Tokens are deposited to the pool
+        (, , uint256[] memory balances, ) = vault.getPoolTokenInfo(address(pool));
+        assertApproxEqAbs(balances[daiIdx], expectedBalances[daiIdx], DELTA, "Pool: Wrong DAI balance");
+        assertApproxEqAbs(balances[usdcIdx], expectedBalances[usdcIdx], DELTA, "Pool: Wrong USDC balance");
+
+        // amountsOut are correct
```
Comment (joaobrunoah) [inline pkg/pool-weighted/test/foundry/WeightedPoolLimits.t.sol:276]:
```suggestion
        // `amountsOut` are correct.
```

File: pkg/pool-weighted/test/foundry/WeightedPoolLimits.t.sol [hunk 8]
```diff
@@ -0,0 +1,383 @@
...
+        // Tokens are stored in the Vault
+        assertApproxEqAbs(
+            usdc.balanceOf(address(vault)),
+            expectedBalances[usdcIdx],
+            DELTA,
+            "Vault: Wrong USDC balance"
+        );
+        assertApproxEqAbs(dai.balanceOf(address(vault)), expectedBalances[daiIdx], DELTA, "Vault: Wrong DAI balance");
+
+        // Tokens are deposited to the pool
```
Comment (joaobrunoah) [inline pkg/pool-weighted/test/foundry/WeightedPoolLimits.t.sol:271]:
```suggestion
        // Tokens are deposited to the pool.
```

File: pkg/pool-weighted/test/foundry/WeightedPoolLimits.t.sol [hunk 9]
```diff
@@ -0,0 +1,383 @@
...
+            TOKEN_AMOUNT,
+            DELTA,
+            "LP: Wrong USDC balance"
+        );
+        assertApproxEqAbs(dai.balanceOf(bob) - startingBalances[daiIdx], TOKEN_AMOUNT, DELTA, "LP: Wrong DAI balance");
+
+        expectedBalances[daiIdx] = TOKEN_AMOUNT - amountsIn[daiIdx];
+        expectedBalances[usdcIdx] = TOKEN_AMOUNT - amountsIn[usdcIdx];
+
+        // Tokens are stored in the Vault
```
Comment (joaobrunoah) [inline pkg/pool-weighted/test/foundry/WeightedPoolLimits.t.sol:262]:
```suggestion
        // Tokens are stored in the Vault.
```

File: pkg/pool-weighted/test/foundry/WeightedPoolLimits.t.sol [hunk 10]
```diff
@@ -0,0 +1,383 @@
...
+        // Tokens are deposited to the pool
+        (, , uint256[] memory balances, ) = vault.getPoolTokenInfo(address(pool));
+        assertApproxEqAbs(balances[daiIdx], expectedBalances[daiIdx], DELTA, "Pool: Wrong DAI balance");
+        assertApproxEqAbs(balances[usdcIdx], expectedBalances[usdcIdx], DELTA, "Pool: Wrong USDC balance");
+
+        // amountsOut are correct
+        assertApproxEqAbs(amountsOut[daiIdx], TOKEN_AMOUNT, DELTA, "Wrong DAI AmountOut");
+        assertApproxEqAbs(amountsOut[usdcIdx], TOKEN_AMOUNT, DELTA, "Wrong USDC AmountOut");
+
+        // should mint correct amount of BPT tokens
```
Comment (joaobrunoah) [inline pkg/pool-weighted/test/foundry/WeightedPoolLimits.t.sol:280]:
```suggestion
        // Should mint correct amount of BPT tokens.
```

File: pkg/pool-weighted/test/foundry/WeightedPoolLimits.t.sol [hunk 11]
```diff
@@ -0,0 +1,383 @@
...
+        startingBalances[daiIdx] = dai.balanceOf(bob);
+        startingBalances[usdcIdx] = usdc.balanceOf(bob);
+
+        vm.prank(bob);
+        uint256 amountCalculated = router.swapSingleTokenExactIn(
+            address(pool),
+            dai,
+            usdc,
+            TOKEN_AMOUNT_IN,
+            0, // don't worry about limit here; will test results anyway
```
Comment (joaobrunoah) [inline pkg/pool-weighted/test/foundry/WeightedPoolLimits.t.sol:297]:
```suggestion
            0, // Don't worry about limit here; will test results anyway.
```

File: pkg/pool-weighted/contracts/test/WeightedPoolMock.sol [hunk 12]
```diff
@@ -0,0 +1,62 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+pragma solidity ^0.8.24;
+
+import {
```
Comment (joaobrunoah) [inline pkg/pool-weighted/contracts/test/WeightedPoolMock.sol:5]: As pointed out by @elshan-eth, there are imported files that are not used.

**Review by joaobrunoah — APPROVED** [review]
LGTM, only small comments.

File: pkg/pool-weighted/test/foundry/WeightedPoolLimits.t.sol (hunk 1 above)
Comment (EndymionJkb) [reply pkg/pool-weighted/test/foundry/WeightedPoolLimits.t.sol:179]: I've trained you well :)

File: pkg/pool-weighted/test/foundry/WeightedPoolLimits.t.sol (hunk 11 above)
Comment (EndymionJkb) [reply pkg/pool-weighted/test/foundry/WeightedPoolLimits.t.sol:297]: Inline comments don't have periods

**Pull request merged**